Yosemite/Yosemite/Stores/NotificationStore.swift

```diff
@@ -218,26 +197,30 @@ extension NotificationStore {
     /// Deletes the collection of local notifications that cannot be found in a given collection of
     /// remote hashes.
     ///
-    /// - Parameter remoteIds: Collection of remote Note IDs.
+    /// - Parameters:
+    ///    - remoteIds: Collection of remote Note IDs.
+    ///    - completion: Callback closure returning outdated note IDs.
     ///
-    func deleteLocalMissingNotes(from hashes: [NoteHash], completion: @escaping (() -> Void)) {
-        let derivedStorage = type(of: self).sharedDerivedStorage(with: storageManager)
-
-        derivedStorage.perform {
+    func deleteLocalMissingNotes(from hashes: [NoteHash], completion: @escaping (([Int64]) -> Void)) {
+        storageManager.performAndSave({ [weak self] storage -> [Int64] in
+            guard let self else { return [] }
             // The beauty of threadsafe Immutable Entities!!
             let remoteIDs = hashes.map { $0.noteID }
             let predicate = NSPredicate(format: "NOT (noteID IN %@)", remoteIDs)
 
-            for orphan in derivedStorage.allObjects(ofType: Storage.Note.self, matching: predicate, sortedBy: nil) {
-                derivedStorage.deleteObject(orphan)
+            let allObjects = storage.allObjects(ofType: Storage.Note.self, matching: predicate, sortedBy: nil)
+            for orphan in allObjects {
+                storage.deleteObject(orphan)
             }
-        }
-
-        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
-            DispatchQueue.main.async {
-                completion()
+            return determineOutdatedNotes(using: hashes, in: storage)
+        }, completion: { result in
+            switch result {
+            case .success(let outdatedNoteIDs):
+                completion(outdatedNoteIDs)
+            case .failure:
+                completion([])
```

Comment:
Since no error is thrown in the operation closure, the failure case should not happen. I added a comment in 7b5027913c68e33cc8a03646764d42b8b49c44d4 to clarify this.